src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -857,27 +860,161 @@ private void InitializeComponentsUsingEnvironment(DocumentDatabase documentDatab
             _indexStorage.Initialize(documentDatabase, environment);
 
             IndexPersistence?.Dispose();
+
             SearchEngineType = IndexStorage.ReadSearchEngineType(Name, environment);
+            IndexFieldsPersistence = new IndexFieldsPersistence(this);
+            IndexFieldsPersistence.Initialize();
 
             switch (SearchEngineType)
             {
                 case SearchEngineType.None:
                 case SearchEngineType.Lucene:
-                    IndexPersistence = new LuceneIndexPersistence(this, documentDatabase.IndexStore.IndexReadOperationFactory);
                     SearchEngineType = SearchEngineType.Lucene;
+                    IndexPersistence = new LuceneIndexPersistence(this, documentDatabase.IndexStore.IndexReadOperationFactory);
+
                     break;
                 case SearchEngineType.Corax:
-                    IndexPersistence = new CoraxIndexPersistence(this, documentDatabase.IndexStore.IndexReadOperationFactory);
                     SearchEngineType = SearchEngineType.Corax;
+                    IndexPersistence = new CoraxIndexPersistence(this, documentDatabase.IndexStore.IndexReadOperationFactory);
                     break;
                 default:
                     throw new InvalidDataException($"Cannot read search engine type for {Name}. Please reset the index.");
             }
 
             IndexPersistence.Initialize(environment);
+            EnsureIsInitialized();
+        }
 
-            IndexFieldsPersistence = new IndexFieldsPersistence(this);
-            IndexFieldsPersistence.Initialize();
+        private void EnsureIsInitialized()
+        {
+            if (SearchEngineType == SearchEngineType.Lucene)
+                return;
+
+            using (_contextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (var tx = context.OpenReadTransaction())
+            {
+                // It is already initialized, we can return.
+                if (CompactTree.HasDictionary(tx.InnerTransaction.LowLevelTransaction))
+                    return;
+            }
+
+            // We know that Corax is not initialized yet, therefore we will call 
+            OnCoraxInitialization();
+        }
+
+        private struct DocumentTrainSpanEnumerator : IReadOnlySpanEnumerator
```

Comment:
Please move it to a dedicated .cs file. Somewhere under `Corax` directory since it's specific only for it